File: examples/06-saga-spring/python/sync_saga_update.py
from dataclasses import dataclass

from temporalio.client import Client, WithStartWorkflowOperation
from temporalio.common import WorkflowIDConflictPolicy

from saga_compensation import OrderSagaWorkflow

# Sync request/response over a saga. execute_update_with_start_workflow creates the
# workflow if it does not exist and applies the update in one round trip — the
# Python equivalent of Java's startUpdateWithStart. The caller waits on the result
# without a separate start RPC.


@dataclass
class OrderRequest:
    order_id: str


async def submit_order(client: Client, request: OrderRequest) -> str:
    start_op = WithStartWorkflowOperation(
        OrderSagaWorkflow.process,
        request.order_id,
        id=f"order-{request.order_id}",
        task_queue="orders",
        id_conflict_policy=WorkflowIDConflictPolicy.USE_EXISTING,
    )

    return await client.execute_update_with_start_workflow(
        "submit",
        request,
        start_workflow_operation=start_op,
    )

File: examples/06-saga-spring/go/sync_saga_update.go
package saga

import (
	"context"

	enumspb "go.temporal.io/api/enums/v1"
	"go.temporal.io/sdk/client"
)

// SubmitOrderSync is sync request/response over a saga. UpdateWithStartWorkflow
// creates the workflow if it does not exist and applies the update in one round
// trip — the Go equivalent of Java's startUpdateWithStart — so the caller can wait
// on a result without a separate start RPC.
func SubmitOrderSync(ctx context.Context, c client.Client, request OrderRequest) (string, error) {
	startOp := c.NewWithStartWorkflowOperation(
		client.StartWorkflowOptions{
			ID:                       "order-" + request.OrderID,
			TaskQueue:                "orders",
			WorkflowIDConflictPolicy: enumspb.WORKFLOW_ID_CONFLICT_POLICY_USE_EXISTING,
		},
		OrderSagaWorkflow,
		request.OrderID,
	)

	handle, err := c.UpdateWithStartWorkflow(ctx, client.UpdateWithStartWorkflowOptions{
		StartWorkflowOperation: startOp,
		UpdateOptions: client.UpdateWorkflowOptions{
			UpdateName:   "submit",
			Args:         []interface{}{request},
			WaitForStage: client.WorkflowUpdateStageCompleted,
		},
	})
	if err != nil {
		return "", err
	}

	var result string
	if err := handle.Get(ctx, &result); err != nil {
		return "", err
	}
	return result, nil
}
